File: 01_arrays/5_sort_list_of_zeroes_ones_twos.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        li = 0
        ri = len(nums) - 1

        curr = li
        while curr <= ri:
            if nums[curr] == 0:
                nums[curr], nums[li] = nums[li], nums[curr]
                li += 1
                curr += 1
            elif nums[curr] == 2:
                nums[curr], nums[ri] = nums[ri], nums[curr]
                ri -= 1
            elif nums[curr] == 1:
                curr += 1

File: 01_arrays/5_sort_list_of_zeroes_ones_twos.rs
impl Solution {
    pub fn sort_colors(nums: &mut Vec<i32>) {
        let n = nums.len();
        let mut l: usize = 0;
        let mut r: usize = n - 1;
        let mut curr: usize = 0;

        while curr <= r {
            if nums[curr] == 0 {
                nums.swap(l, curr);
                l += 1;
                curr += 1;
            } else if nums[curr] == 2 {
                nums.swap(r, curr);
                if r == 0 {
                    break;
                }
                r -= 1;
            } else {
                curr += 1
            }
        }
    }
}
